src/shell/rexec.c

```diff
@@ -0,0 +1,136 @@
+/************************************************************\
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* rexec.c - shell subprocess server
+ */
+
+#define FLUX_SHELL_PLUGIN_NAME "rexec"
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <sys/types.h>
+#include <signal.h>
+#include <unistd.h>
+
+#include <flux/core.h>
+#include <jansson.h>
+
+#include "src/common/libsubprocess/server.h"
+#include "src/common/libutil/errprintf.h"
+
+#include "builtins.h"
+#include "internal.h"
+#include "svc.h"
+#include "log.h"
+
+struct shell_rexec {
+    flux_shell_t *shell;
+    subprocess_server_t *server;
+    char *name;
+    bool parent_is_trusted;
+};
+
+static void rexec_destroy (struct shell_rexec *rexec)
+{
+    if (rexec) {
+        int saved_errno = errno;
+        subprocess_server_destroy (rexec->server);
+        free (rexec->name);
+        free (rexec);
+        errno = saved_errno;
+    }
+}
+
+/* The embedded subprocess server restricts access based on FLUX_ROLE_OWNER,
+ * but this shell cannot trust message credentials if they are passing through
+ * a Flux instance running as a different user (e.g. the "flux" user in a
+ * system instance).  If that user were compromised, they could run arbitrary
+ * commands as any user that currently has a job running.  Therefore, this
+ * additional check ensures that we only trust an instance running as the same
+ * user.
+ */
+static int rexec_auth_cb (const flux_msg_t *msg,
+                          void *arg,
+                          flux_error_t *errp)
+{
+    struct shell_rexec *rexec = arg;
+
+    if (!rexec->parent_is_trusted) {
+        errno = EPERM;
+        return errprintf (errp, "Access denied");
+    }
```

Comment:
Is the normal uid auth check still necessary here? That is, would this prevent a job running as user `flux` in a system instance from allowing an arbitrary user to exec something as user `flux` within that job?

Reply:
Should be OK.  An earlier version of this PR included a commit that allowed the subprocess server message handlers to be registered with an arbitrary rolemask, but I reverted that and so now they are registered for owner only.  I think the combination of that check and the one quoted should be sufficient.